staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go

```diff
@@ -697,6 +705,60 @@ func (s *Scheme) addObservedVersion(version schema.GroupVersion) {
 	s.observedVersions = append(s.observedVersions, version)
 }
 
+func objectAPILifecycle(obj Object) (schema.APILifecycle, bool) {
+	hasLifecycle := false
+	ret := schema.APILifecycle{}
+	introduced, isIntroduced := obj.(apiLifecycleIntroduced)
+	if isIntroduced {
+		major, minor := introduced.APILifecycleIntroduced()
+		ret.IntroducedVersion = version.MajorMinor(uint(major), uint(minor))
+		hasLifecycle = true
+	}
+	removed, isRemoved := obj.(apiLifecycleRemoved)
+	if isRemoved {
+		major, minor := removed.APILifecycleRemoved()
+		ret.RemovedVersion = version.MajorMinor(uint(major), uint(minor))
+		hasLifecycle = true
+	}
+
+	return ret, hasLifecycle
+}
+
+type apiLifecycleIntroduced interface {
+	APILifecycleIntroduced() (major, minor int)
+}
+
+type apiLifecycleRemoved interface {
+	APILifecycleRemoved() (major, minor int)
+}
+
+// SetGroupVersionLifecycle sets the APILifecycle for the GroupVersion.
+func (s *Scheme) SetGroupVersionLifecycle(gv schema.GroupVersion, lifecycle schema.APILifecycle) {
+	s.gvToAPILifecycle[gv] = lifecycle
+}
+
+// SetResourceLifecycle sets the APILifecycle for the GroupVersionResource.
+func (s *Scheme) SetResourceLifecycle(gvr schema.GroupVersionResource, obj Object) {
+	lifecycle, hasLifecycle := objectAPILifecycle(obj)
+	if hasLifecycle {
+		s.gvrToAPILifecycle[gvr] = lifecycle
+	}
+}
+
+// GroupVersionLifecycle returns the APILifecycle for the GroupVersion.
+func (s *Scheme) GroupVersionLifecycle(gv schema.GroupVersion) schema.APILifecycle {
+	return s.gvToAPILifecycle[gv]
+}
+
+// ResourceLifecycle returns the APILifecycle for the GroupVersionResource.
+func (s *Scheme) ResourceLifecycle(gvr schema.GroupVersionResource) schema.APILifecycle {
+	if lifecycle, ok := s.gvrToAPILifecycle[gvr]; ok {
+		return lifecycle
+	}
+	// fallback to GroupVersionLifecycle if ResourceLifecycle is not set.
+	return s.GroupVersionLifecycle(gvr.GroupVersion())
```

Comment:
Ah, great, this is a fallback. Document that this fallback happens in the godoc of the functions that set lifecycles?

Reply:
Also, if the GVR only provides a removed version, should the backfill kick in for the introduced version? Rationale and docs around this seems valuable.